chronos/asyncproc.nim

```diff
@@ -0,0 +1,1323 @@
+      # Make a duplicate of `stdoutHandle` as child process STDOUT.
+      doCheck(posixSpawnFileActionsAddDup2(actions, cint(pipes.stdoutHandle),
+                                           cint(1)))
+      if AsyncProcessOption.StdErrToStdOut notin options:
+        # Close child process side of `stdoutHandle`.
+        doCheck(posixSpawnFileActionsAddClose(actions,
+                                              cint(pipes.stdoutHandle)))
+        # Close parent process side of `stdoutHandle`.
+        if not(pipes.stdoutHolder.isEmpty()):
+          let fd = cint(pipes.stdoutHolder.fd())
+          doCheck(posixSpawnFileActionsAddClose(actions, fd))
+
+    if pipes.flags * {ProcessFlag.AutoStderr, ProcessFlag.UserStderr} != {}:
+      # Close child process STDERR.
+      doCheck(posixSpawnFileActionsAddClose(actions, cint(2)))
+      # Make a duplicate of `stderrHandle` as child process STDERR.
+      doCheck(posixSpawnFileActionsAddDup2(actions, cint(pipes.stderrHandle),
+                                           cint(2)))
+      # Close child process side of `stderrHandle`.
+      doCheck(posixSpawnFileActionsAddClose(actions,
+                                            cint(pipes.stderrHandle)))
+      # Close parent process side of `stderrHandle`.
+      if not(pipes.stderrHolder.isEmpty()):
+        let fd = cint(pipes.stderrHolder.fd())
+        doCheck(posixSpawnFileActionsAddClose(actions, fd))
+    else:
+      if AsyncProcessOption.StdErrToStdOut in options:
+        # Close child process STDERR.
+        doCheck(posixSpawnFileActionsAddClose(actions, cint(2)))
+        # Make a duplicate of `stdoutHandle` as child process STDERR.
+        doCheck(posixSpawnFileActionsAddDup2(actions, cint(pipes.stdoutHandle),
+                                             cint(2)))
+        # Close child process side of `stdoutHandle`.
+        doCheck(posixSpawnFileActionsAddClose(actions,
+                                              cint(pipes.stdoutHandle)))
+        # Close parent process side of `stdoutHandle`.
+        if not(pipes.stdoutHolder.isEmpty()):
+          let fd = cint(pipes.stdoutHolder.fd())
+          doCheck(posixSpawnFileActionsAddClose(actions, fd))
+    ok(SpawnAttr(attrs: attrs, actions: actions))
+
+  proc free(v: var SpawnAttr): Result[void, OSErrorCode] =
+    block:
+      let res = posixSpawnAttrDestroy(v.attrs)
+      if res != 0:
+        discard posixSpawnFileActionsDestroy(v.actions)
+        return err(OSErrorCode(res))
+    block:
+      let res = posixSpawnFileActionsDestroy(v.actions)
+      if res != 0:
+        return err(OSErrorCode(res))
+    ok()
+
+  proc envToCStringArray(t: StringTableRef): cstringArray =
+    let itemsCount = len(t)
+    var
+      res = cast[cstringArray](alloc0((itemsCount + 1) * sizeof(cstring)))
+      i = 0
+    for key, value in pairs(t):
+      var x = key & "=" & value
+      res[i] = cast[cstring](alloc0(len(x) + 1))
```

Comment:
The zero-initialization of `alloc0` is unnecessary here. All bytes will be overwritten.

Reply:
done